### DISCUSSION TITLE
Switching  to I2S  and  A2DP  when blutooth  speaker  no  longer  exist

### DISCUSSION THREAD

**Discussion by vednildip** [pschatzmann/arduino-audio-tools#1832] [Q&A] [answered]
A2DP StreamCopy copier   implementation  has   limitation  A2DPcopier.copy   that  even  Bluetooth  speaker  does  not  exist  for  longer  , the   function  loops infinite  connecting  &  disconnecting  cycle so  coming  out  of  that  loop  becomes  difficult  ,  isConnected   gives  value   0    but   can  not  use  to  connect  I2Scopier.copy  . Used  millis() logic, setActive flag ,   but  failed . if  any  one    has  any  recomendation  please  inform  me

**Comment by pschatzmann** [comment, marked as answer]
I have no clue what you talk about! 
Neither A2DPcopier nor I2Scopier is anything I have ever implemented!

Try to describe what you are doing and what your problem is in a way that can be understood and reproduced by others!

**Reply by vednildip** [reply]
StreamCopy copier(a2dpStream, i2sStream);   &  StreamCopy copier(i2s, i2sStream); 
 instead  just copier  I had  created  object  A2DPcopier &   I2Scopier.  Object  can  be  of  intuitive  name  if  I  am  using  both  in  one  program  one  after  another  calling  if    Bluetooth speaker does not exist for longer

**Comment by pschatzmann** [comment]
This is getting even more confusing:  I still don't understand what you try to do and what your problem is!
Try to describe what you are doing and what your problem is in a way that can be understood and reproduced by others!

**Comment by vednildip** [comment]
Input ( RX)  received   ESP32 Node MCU  copying  for  output   either  to  Bluetooth  speaker(a2dpStream)   or  directly  to  i2S  wired to speaker(i2s) .  Both  work  individual  call copier.copy () calling  function.  implementing  logic  if  Bluetooth  speaker disconnected    copier.copy ()  try  to  connect for  infinite  loop   want  to  break  that  loop  for  example  after  10 second  and   switch  it  to     directly  to  i2S  wired to speaker(i2s) .

**Comment by pschatzmann** [comment]
Just don't call copy() if a2dp is not connected!
You could work with one single StreamCopy to reflect your logic that it is one or the other and make sure that you use the correct destination  dependent on the a2dp connection state.

**Comment by vednildip** [comment]
this is okay  first  time  , but  in  between  Bluetooth speaker  battery  get  exhausted, then    copier.copy  does  not come  out  from  connecting  , disconnecting  loop  is  limitation

**Comment by pschatzmann** [comment]
I have added the tx_write_timeout_ms to the A2DPConfig. By default (with value -1) it is blocking endlessly. 
So just define a timout value...

**Reply by vednildip** [reply]
I  am  very much  happy  as code run as per expectation what  I  intend  to  experiment . 

 Quote  for  you  "What you gain or earn in this life is not what truly matters. It is the transformative changes you bring to society at large that will be remembered by history. "
